### PR TITLE
Rust SDK for data transforms

### PR DESCRIPTION
**Pull request by kbatuigas**
Preview pages:
- [Reference](https://deploy-preview-279--redpanda-docs-preview.netlify.app/current/reference/data-transform-rust-sdk/)
- [Tutorial](https://deploy-preview-279--redpanda-docs-preview.netlify.app/current/develop/data-transforms/run-transforms/)

This PR:
- Adds Rust SDK reference doc
- Adds Rust example to [data transforms tutorial](https://docs.redpanda.com/current/develop/data-transforms/run-transforms/)
- Also changes the nav tree section to "API and SDK reference" to better describe the contents in the section. The title of the Go SDK page has also been updated for consistency with the new Rust doc, and make it easier overall for readers to navigate to / find the reference doc they want

Closes https://github.com/redpanda-data/documentation-private/issues/2164

### CODE REVIEW

File: modules/develop/partials/run-transforms.adoc
Comment (kbatuigas) [inline modules/develop/partials/run-transforms.adoc]: Ok to remove copyright block of comments?

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (kbatuigas) [inline modules/reference/pages/data-transform-rust-sdk.adoc]: Generated doc still says "trait" https://docs.rs/redpanda-transform-sdk/0.2.0/redpanda_transform_sdk/struct.RecordWriter.html Should the description be updated? Also, I'm finding "for output transformed records to the destination topic" hard to understand, can this be rephrased or is there just an incorrect word here?

**Comment by kbatuigas** [comment]
@rockwotj Could we add the following details for Rust to the tutorial?

* Any prereqs for Rust: https://deploy-preview-279--redpanda-docs-preview.netlify.app/current/develop/data-transforms/run-transforms/#prerequisites
* Runtime overhead https://deploy-preview-279--redpanda-docs-preview.netlify.app/current/develop/data-transforms/run-transforms/#configure-memory
* This section is written entirely for Go - could we get an example file structure for Rust? https://deploy-preview-279--redpanda-docs-preview.netlify.app/current/develop/data-transforms/run-transforms/#create-a-data-transforms-project 
** Is it enough to just swap out the Go filenames to the Rust equivalents? What about the information in the [TIP]?
* Do the rpk commands need the language/framework to be specified?

**Comment by rockwotj** [comment]
> Any prereqs for Rust: https://deploy-preview-279--redpanda-docs-preview.netlify.app/current/develop/data-transforms/run-transforms/#prerequisites

You can replace `At least version 1.20 of [Go](https://go.dev/doc/install) installed on your host machine.` with `The latest stable version of Rust installed on the host machine via [rustup](https://rustup.rs/).`

> Runtime overhead https://deploy-preview-279--redpanda-docs-preview.netlify.app/current/develop/data-transforms/run-transforms/#configure-memory

We should remove the lines about overhead - we don't support standard Golang right now and Tinygo + Rust's memory overhead is not significant.

> This section is written entirely for Go - could we get an example file structure for Rust? https://deploy-preview-279--redpanda-docs-preview.netlify.app/current/develop/data-transforms/run-transforms/#create-a-data-transforms-project ** Is it enough to just swap out the Go filenames to the Rust equivalents? 

Yes, the rust equivalents are now merged into `dev` here: https://github.com/redpanda-data/redpanda/tree/dev/src/transform-sdk/rust/examples

> What about the information in the [TIP]?

Everything there is still valid except for this bit: `or in Go’s standard predefined init() function`.

> Do the rpk commands need the language/framework to be specified?

`rpk transform init` will ask the user if they don't specify via `--language=<lang>`, which is good. The file layout for rust will be different:

```
.
├── Cargo.lock
├── Cargo.toml
├── README.md
├── src
│  └── main.rs
└── transform.yaml
```

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (rockwotj) [reply modules/reference/pages/data-transform-rust-sdk.adoc]: Ah thanks, yes it's a struct now, I missed that - will send out a PR.

> "for output transformed records to the destination topic" hard to understand, can this be rephrased or is there just an incorrect word here.

Yes needs rephrased: It's a struct for writing transformed records to the output topic.

File: modules/develop/partials/run-transforms.adoc
Comment (rockwotj) [reply modules/develop/partials/run-transforms.adoc]: Yes

File: modules/develop/partials/run-transforms.adoc
Comment (kbatuigas) [inline modules/develop/partials/run-transforms.adoc:23]: @rockwotj Is TinyGo still supported? If so, I'll add a link to the [install instructions](https://tinygo.org/getting-started/install/).

The `rpk transform init` command further below also still shows the Go project files, I'm assuming the flag can be either `rust` or `go`?

File: modules/develop/pages/data-transforms/how-transforms-work.adoc
Comment (rockwotj) [inline modules/develop/pages/data-transforms/how-transforms-work.adoc:49]: We should probably have the go sdk here too right?

File: modules/develop/partials/run-transforms.adoc
Comment (rockwotj) [reply modules/develop/partials/run-transforms.adoc:23]: They don't need to install TinyGo, just standard Go: https://go.dev/doc/install

This line is needed for Golang development

If you want to do Rust development you need rustup installed (but not Golang).

File: modules/develop/partials/run-transforms.adoc
Comment (rockwotj) [inline modules/develop/partials/run-transforms.adoc]: This is `main.rs` if rust (same with the next paragraph).

File: modules/develop/partials/run-transforms.adoc
Comment (rockwotj) [inline modules/develop/partials/run-transforms.adoc]: These read very different:

Screenshot 2024-01-30 at 8 40 12 AM

Should we unify the names so it is clear they are the same thing for different languages?

File: modules/develop/partials/run-transforms.adoc
Comment (rockwotj) [inline modules/develop/partials/run-transforms.adoc]: Now that we are going to have these per language, should this be `data-transform-golang-sdk`?

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (rockwotj) [inline modules/reference/pages/data-transform-rust-sdk.adoc:1]: I'm not seeing this page in the left nav. Am I missing something?

**Review by rockwotj — COMMENTED** [review]
Looking good. A few things I noticed

File: modules/develop/partials/run-transforms.adoc
Comment (rockwotj) [inline modules/develop/partials/run-transforms.adoc:106]: Does it make sense to have a single tab for this whole section instead of two tabs? I don't know how that looks, but something to consider.

**Review by rockwotj — APPROVED** [review]
I have a single question about if consolidating the tabs is a good idea or not, but otherwise this is awesome thank you for your work here!

File: modules/develop/partials/run-transforms.adoc
Comment (JakeSCahill) [inline modules/develop/partials/run-transforms.adoc]:
```suggestion
- Development tools installed on your host machine:
** For Golang, you must have at least version 1.20 of https://go.dev/doc/install[Go^].
** For Rust, you must have the latest stable version of https://rustup.rs/[Rust].
```

See https://github.com/redpanda-data/docs-site/blob/main/meta-docs/STYLE-GUIDE.adoc#:~:text=use%20Shadow%20Indexing.-,via,-Do%20not%20use

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]: Is this heading required? we could remove this heading and bump the others to avoid too much nesting.

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]: Is it worth having a short intro to describe a struct with a link to the Rust docs?

https://doc.rust-lang.org/rust-by-example/custom_types/structs.html

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]: maybe a short intro to describe enums with a link to the Rust doc: https://doc.rust-lang.org/rust-by-example/custom_types/enum.html

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]: same here - a short intro would be helpful to set some context with a link for those who aren't familiar: https://doc.rust-lang.org/rust-by-example/trait.html

Comment (JakeSCahill) [reply modules/reference/pages/data-transform-rust-sdk.adoc]: Also, traits aren't types so would be better to make this the same level as 'Data transforms types'

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]: short intro to functions and a link: https://doc.rust-lang.org/rust-by-example/fn.html

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]: maybe a short intro to how the SDK provides custom types such as structs and enums designed to handle data efficiently, ensuring high performance and ease of use.

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]:
```suggestion
The `on_record_written()` function transforms individual records after they have been written to an input topic. The function registers a callback to be fired when a record is written to the topic. This callback is triggered after the record has been written and fsynced to disk and the producer has been acknowledged.
```

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]:
```suggestion
Since this method blocks and runs forever, it should be called from the `main()` function. Perform the necessary steps before transforming any records before calling this method.
```

Comment (JakeSCahill) [reply modules/reference/pages/data-transform-rust-sdk.adoc]: Google style guide suggests using `()` after function/method names to make it clearer that they are executable: https://developers.google.com/style/code-in-text#method-names

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]:
```suggestion
`BorrowedHeader` is a zero-copy <<borrowedrecord,`BorrowedRecord`>> header.
```

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]:
```suggestion
`BorrowedRecord` is a zero-copy representation of a <<record,`Record`>> in Redpanda. 
```

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]:
```suggestion
`Record` is a record in Redpanda. It consists of a key-value pair of bytes, along with a collection of <<recordheader,`RecordHeader`>>.
```

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]:
```suggestion
A `RecordHeader` is a key-value pair attached to a <<record,`Record`>>. Headers are opaque to the broker and are purely a mechanism for the producer and consumers to pass information.
```

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]: Isn't this passed to `on_record_written()`?

```rust
pub fn on_record_written<E, F>(cb: F) -> !
where
    E: Debug,
    F: Fn(WriteEvent<'_>, &mut dyn RecordWriter) -> Result<(), E>,
```

**Review by JakeSCahill — COMMENTED** [review]
Some suggestions and formatting fixes.

I've also added a PR to adjust the UI so that we don't try to replace Rust syntax with editable placeholders. https://github.com/redpanda-data/docs-ui/pull/148

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (kbatuigas) [reply modules/reference/pages/data-transform-rust-sdk.adoc]: @rockwotj Is there a suitable heading for traits that would match the others at this level? (i.e. `Data transforms functions`, 'Data transforms types`)

Comment (rockwotj) [reply modules/reference/pages/data-transform-rust-sdk.adoc]: > Also, traits aren't types so would be better to make this the same level as 'Data transforms types'

structs, enums and traits are all kinds of types and are all involved in the typesystem, but different entities I think it is valid to do one of the following approaches:

* Have headings for `trait`, `struct`, `enum`, `function`
(this is what [docs.rs](https://docs.rs/redpanda-transform-sdk/0.2.0/redpanda_transform_sdk/) does so I lean that way)

* The more [go style](https://pkg.go.dev/github.com/redpanda-data/redpanda/src/transform-sdk/go/transform) that has two groups: `types` and `functions`.

I don't have a strong preference. If you want to do the two level heading like golang, that's fine with me.

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (rockwotj) [reply modules/reference/pages/data-transform-rust-sdk.adoc]: Yes WriteEvents are. I am trying to communicate they these events are being processed after the broker has already completed the write.

File: modules/develop/partials/run-transforms.adoc
Comment (rockwotj) [reply modules/develop/partials/run-transforms.adoc:106]: Thanks I really actually like the single tab here!

File: modules/develop/partials/run-transforms.adoc
Comment (rockwotj) [inline modules/develop/partials/run-transforms.adoc]:
```suggestion
TIP: When creating a custom data transform, initialization steps can be done in `main()` because it's only run once at the start of the package. Although state can be cached in global variables, Redpanda may restart a Wasm module at any point, which causes the state to be lost.
```

It's the main function, not really just the main file.

**Review by rockwotj — APPROVED** [review]
Just one thing I caught on a second review

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]: Not sure what this sentence should say but seems like a typo:

> Perform the necessary steps **before** transforming any records **before** calling this method.

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]:
```suggestion
The `on_record_written()` function registers a callback that transforms records after they have been written to an input topic. This function is executed when a record is written to the topic. The callback is executed after the record has been written and fsynced to disk and the producer has been acknowledged.

Callback parameters

- <<writeevent,`WriteEvent`>>
- <<recordwriter, `RecordWriter`>>

Example:

This transform function copies the same data from an input topic to an output topic.:

[,rust]
----
use anyhow::Result;
use redpanda_transform_sdk::*;

fn main() {
    on_record_written(my_transform);
}

// This will be called for each record in the source topic. The output records returned will be written to the destination topic.
fn my_transform(event: WriteEvent, writer: &mut RecordWriter) -> Result<()> {
    writer.write(event.record)?;
    Ok(())
----
```

Maybe we should detail what happens to errors? Does the error interrupt the writing process?

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc:188]: how come we don't document the properties and methods for these structs. How should users know that `writer.record` or `writer.record.key()` exists?

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [inline modules/reference/pages/data-transform-rust-sdk.adoc]:
```suggestion
A https://doc.rust-lang.org/rust-by-example/trait.html[trait] defines behavior that a type shares with other types.
```

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [reply modules/reference/pages/data-transform-rust-sdk.adoc]: I'd prefer to start this paragraph with the action required followed by the reason:

```
NOTE: You must call this method from the `main()` function. REASON...
```

**Review by JakeSCahill — COMMENTED** [review]
On a second review, I noticed a few things that weren't clear. Would be good to clarify these before we merge.

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (JakeSCahill) [reply modules/reference/pages/data-transform-rust-sdk.adoc]: Might also be nice to show error handling in the example

Comment (kbatuigas) [reply modules/reference/pages/data-transform-rust-sdk.adoc]: @rockwotj Would you be able to provide more details on these?

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (kbatuigas) [reply modules/reference/pages/data-transform-rust-sdk.adoc:188]: May have just missed this info, @rockwotj would you mind confirming if this has all been provided in https://docs.rs/redpanda-transform-sdk/0.2.0/redpanda_transform_sdk/index.html?

Comment (rockwotj) [reply modules/reference/pages/data-transform-rust-sdk.adoc:188]: Oh yes silly me. For each of these times we still need the methods for each struct/trait:

For RecordWriter:

https://docs.rs/redpanda-transform-sdk/0.2.0/redpanda_transform_sdk/struct.RecordWriter.html

You'll have to drill down to get more information.

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (rockwotj) [reply modules/reference/pages/data-transform-rust-sdk.adoc]: There is "error handling" in that if `write` fails we return an error. That's what the `?` does. I don't think it's expected that users handle errors in this example.

If an error is returned, we end up printing an error and aborting the "process" in the VM, and then Redpanda will retry processing with backoff.

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (rockwotj) [inline modules/reference/pages/data-transform-rust-sdk.adoc:45]: missing closing `}` on the next line.

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (rockwotj) [inline modules/reference/pages/data-transform-rust-sdk.adoc]: Let's remove the insides of these, they are likely to go out of date and they are private.

File: modules/reference/pages/data-transform-rust-sdk.adoc
Comment (rockwotj) [inline modules/reference/pages/data-transform-rust-sdk.adoc]: Can we just remove everything inside the `{ ... }` for these? The signature is the only important thing here. You can just replace them with a tailing semicolon like in the `trait` we have.

Comment (kbatuigas) [reply modules/reference/pages/data-transform-rust-sdk.adoc]: @rockwotj is this correct?
```
    pub fn new(key: &'a [u8], value: Option<&'a [u8]>) -> Self;
```

Comment (rockwotj) [reply modules/reference/pages/data-transform-rust-sdk.adoc]: Yes

File: modules/develop/partials/run-transforms.adoc
Comment (rockwotj) [inline modules/develop/partials/run-transforms.adoc:304]: Missing a trailing `}`

**Review by rockwotj — APPROVED** [review]
One small bit for the identity transform in the tutorial page, otherwise looks amazing. Thank you! 🙏

**Review by rockwotj — APPROVED** [review]

**Review by JakeSCahill — APPROVED** [review]
Very cool - thanks Kat 🙌

**Pull request merged**
